airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```
@@ -0,0 +1,345 @@
+        # minutes. But in case reports wont be generated we dont want this stream to
+        # hung forever. Store timepoint when report generation has started to
+        # check if it takes to long to break a loop.
+        start_time_point = datetime.now()
+        while report_infos and datetime.now() <= start_time_point + self.REPORT_WAIT_TIMEOUT:
+            completed_reports = []
+            logger.info(f"Checking report status, {len(report_infos)} report(s) remained")
+            for report_info in report_infos:
+                report_status, download_url = self._check_status(report_info)
+                if report_status == Status.FAILURE:
+                    logger.warn(f"Report for {report_info.profile_id} with {report_info.record_type} type generation failed")
+                    completed_reports.append(report_info)
+                elif report_status == Status.SUCCESS:
+                    metric_objects = self._download_report(report_info, download_url)
+                    for metric_object in metric_objects:
+                        yield self._model(
+                            profileId=report_info.profile_id,
+                            recordType=report_info.record_type,
+                            reportDate=report_date,
+                            metric=metric_object,
+                        ).dict()
+                    completed_reports.append(report_info)
+            for completed_report in completed_reports:
+                report_infos.remove(completed_report)
+            if report_infos:
+                logger.info(f"{len(report_infos)} report(s) remained, taking {self.CHECK_INTERVAL_SECONDS} seconds timeout")
+                time.sleep(self.CHECK_INTERVAL_SECONDS)
+        if not report_infos:
+            logger.info("All reports have been processed")
+        else:
+            logger.error("Not all reports has been processed due to timeout")
```

we should raise an exception if some reports are not finished

---

Done